apps/xprof_nif/xprof_core_nif_tracer.c

```diff
@@ -0,0 +1,259 @@
+        {"enabled", 3, enabled},
+        {"trace", 5, trace},
+        {"trace_call", 5, trace_call}
+};
+
+ERL_NIF_INIT(xprof_core_nif_tracer, nif_funcs, load, NULL, upgrade, unload)
+
+/* Local functions */
+
+static void read_mfa(ErlNifEnv * env,
+                     const ERL_NIF_TERM mfa_tuple,
+                     char * module,
+                     char * function,
+                     int  * arity)
+{
+    const ERL_NIF_TERM * tuple;
+    int                  tuple_arity;
+
+    enif_get_tuple(env, mfa_tuple, &tuple_arity, &tuple);
+    enif_get_atom(env, tuple[0], module,   256, ERL_NIF_LATIN1);
+    enif_get_atom(env, tuple[1], function, 256, ERL_NIF_LATIN1);
+    enif_get_int(env, tuple[2], arity);
+}
+
+static ERL_NIF_TERM create_mfa_tuple(ErlNifEnv * env,
+                             char * module,
+                             char * function,
+                             int    arity)
+{
+    return enif_make_tuple3(
+            env,
+            enif_make_atom(env, module),
+            enif_make_atom(env, function),
+            enif_make_int (env, arity)
+    );
+}
+
+static ERL_NIF_TERM mfa_to_bin(ErlNifEnv * env,
+                               const ERL_NIF_TERM mfa_tuple)
+{
+    const ERL_NIF_TERM * tuple;
+    int                  tuple_arity;
+
+    unsigned char module[256];
+    unsigned char function[256];
+    int           arity;
+
+    enif_get_tuple(env, mfa_tuple, &tuple_arity, &tuple);
+    enif_get_atom(env, tuple[0], module,   256, ERL_NIF_LATIN1);
+    enif_get_atom(env, tuple[1], function, 256, ERL_NIF_LATIN1);
+    enif_get_int(env, tuple[2], &arity);
+
+    unsigned char buffer[256];
+    sprintf(buffer, "%s:%s/%d", module, function, arity);
+
+    int len = strlen(buffer);
+    ERL_NIF_TERM bin;
+    unsigned char * binstr = enif_make_new_binary(env, len, &bin);
+    memcpy(binstr, buffer, len);
+
+    return bin;
+}
+
+static ERL_NIF_TERM create_timestamp(ErlNifEnv * env)
+{
+    struct timeval timepoint;
+    gettimeofday(&timepoint, NULL);
+
+    int megasec  = timepoint.tv_usec / 1000000000000;
+    int sec      = timepoint.tv_usec / 1000000 - megasec * 1000000;
+    int microsec = timepoint.tv_usec % 1000000;
+
+    return enif_make_tuple3(
+            env,
+            enif_make_int(env, megasec),
+            enif_make_int(env, sec),
+            enif_make_int(env, microsec)
+    );
+}
+
+/* End of locals */
+
+static int load(ErlNifEnv* env, void** priv_data, ERL_NIF_TERM load_info)
+{
+    *priv_data = NULL;
+    return 0;
+}
+
+static void unload(ErlNifEnv* env, void* priv_data)
+{
+
+}
+
+static int upgrade(ErlNifEnv* env, void** priv_data, void** old_priv_data,
+                   ERL_NIF_TERM load_info)
+{
+    if (*old_priv_data != NULL || *priv_data != NULL) {
+        return -1; /* Don't know how to do that */
+    }
+    if (load(env, priv_data, load_info)) {
+        return -1;
+    }
+    return 0;
+}
+
+/*
+ * argv[0]: TraceTag
+ * argv[1]: TracerState
+ * argv[2]: Tracee

```

Comment:
instead of documenting it here consider assigning these values to variables for example:
```
const ERL_NIF_TERM TraceTag = argv[0];
```
Having this will make reading the code below easier.

`enif_is_identical(enif_make_atom(env, "trace_status"), argv[0]))` 
=> 
`enif_is_identical(enif_make_atom(env, "trace_status"), TraceTag)`

Reply:
good idea, nice convention